packages/api/src/apps/api/app.ts

```diff
@@ -233,17 +249,65 @@ export default async (env: ProcessEnv) => {
 
   console.log("Starting API update loops");
 
+  // listen for new lsp contracts since after we have started api, and make sure they get state updated asap
+  // These events should only be bound after startup, since initialization above takes care of updating all contracts on startup
+  // Because there is now a event driven depdency, the lsp creator and lsp state updater must be in same process
+  serviceEvents.on("multiLspCreator", (data) => {
+    console.log("LspCreator found a new contract", JSON.stringify(data));
+    services.lsps.updateLsp(data.address, data.startBlock, data.endBlock);
+  });
+
+  // listen for new emp contracts since after we start api, make sure they get state updates asap
+  serviceEvents.on("empRegistry", (data) => {
+    console.log("EmpRegistry found a new contract", JSON.stringify(data));
+    services.emps.updateOne(data.address, data.startBlock, data.endBlock);
+  });
+
+  async function getLatestBlockNumber() {
+    const block = await provider.getBlock("latest");
+    return block.number;
+  }
+
+  const newContractBlockTick = BlockInterval(detectNewContracts, initBlock.number);
+  const updateContractStateTick = BlockInterval(updateContractState, initBlock.number);
+
+  // detect contract loop
+  utils.loop(async () => {
+    const end = profile("Detecting New Contracts");
+    // adding in a timeout rejection if the update takes too long
```

Comment:
It seems like we've had a few issues with hanging throughout the API application? Do you think there's a root cause for all of these problems? It seems a little strange that we're hanging rather than seeing some sort of error. Are we ignoring errors somewhere? Do we have promises where we only provide a `.then` (with a single arg) and no `.catch` so the promise just remains unresolved forever?

Reply:
The symptom im seeing and concluding things are hanging, is that i've added a specific log for when state updates run. After the servers are up for some time, i stop seeing that log, though price updates still run ( they are in a separate loop). I've been pretty careful around making sure we have visibility around errors, so i would expect to see them if there was one.  If there are any spots where we don't catch an error and one is thrown, it would cause the api to crash, or at least log that theres an uncaught promise error, but logs dont show this. 

The one place where funny business can happen though is around the websocket connection to ethers. Last PR updated the websocket options for web3, but ethers under the hood is using an infura websocket connection. If for some reason the websocket gets disconnected, this is the most likely place where we could just hang forever in a promise without throwing an error, we make tons of requests, then get disconnected before all are resolved ( and ethers swallows the error).  

I didnt want to dig into the guts of this, but there are events on the websocket to listen for disconnections. I just wanted to be sure that we are really getting blocked here first.  If thats the case it would warrant a deeper look into the websocket.

Reply:
Are you sure the websocket hangs on disconnect? Whenever we've had a websocket disconnect in the bots, we get a bunch of throws due to contract calls failing. Is there a way we can verify that the websocket is the source of the hanging as opposed to something that's more directly under our control?

Reply:
Either way, this architectural change seems good to me! The hanging issue seems to keep coming up, though, so my hope is that we can dig into the source of the issue so we don't keep seeing related issues.

Reply:
just to be clear here, the type of blocking im talking about for this issue is different from the api server completely locking up (which actually hasn't happened in weeks now).  This issue im seeing is that a particular update loop just stops looping and therefore contracts are not being updated/discovered. 

> Are you sure the websocket hangs on disconnect? Whenever we've had a websocket disconnect in the bots, we get a bunch of throws due to contract calls failing.

I am totally guessing as to what might be causing this, but the possibility remains open that this is a websocket issue, because most processes have been using web3, not ethers, so its possible there is different behavior from what we have seen in the past. 

>  Is there a way we can verify that the websocket is the source of the hanging as opposed to something that's more directly under our control?

Part of this pr is kind of separating out possible problem areas of code to narrow down where the issue is. The culprit is most likely in the emp-state/lsp-state update services, and the major dependency there is ethers and infura. There is more work i can do to dig into the websocket but would like to see first if any loops time out.